src/platform_impl/web/web_sys/mod.rs

```diff
@@ -52,38 +52,51 @@ impl WindowExtWebSys for Window {
     }
 }
 
-pub fn window_size(window: &web_sys::Window) -> LogicalSize<f64> {
-    let width = window
-        .inner_width()
-        .expect("Failed to get width")
-        .as_f64()
-        .expect("Failed to get width as f64");
-    let height = window
-        .inner_height()
-        .expect("Failed to get height")
-        .as_f64()
-        .expect("Failed to get height as f64");
-
-    LogicalSize { width, height }
-}
-
 pub fn scale_factor(window: &web_sys::Window) -> f64 {
     window.device_pixel_ratio()
 }
 
-pub fn set_canvas_size(canvas: &Canvas, new_size: Size) {
-    let scale_factor = scale_factor(canvas.window());
+pub fn set_canvas_size(
+    window: &web_sys::Window,
+    raw: &HtmlCanvasElement,
+    mut new_size: LogicalSize<f64>,
+) {
+    let document = window.document().expect("Failed to obtain document");
+
+    if !document.contains(Some(raw)) {
+        return;
+    }
+
+    let style = window
+        .get_computed_style(raw)
+        .expect("Failed to obtain computed style")
+        // this can't fail: we aren't using a pseudo-element
+        .expect("Invalid pseudo-element");
+    if style.get_property_value("box-sizing").unwrap() == "border-box" {
+        new_size.width += style_size_property(&style, "border-left-width")
+            + style_size_property(&style, "border-right-width")
+            + style_size_property(&style, "padding-left")
+            + style_size_property(&style, "padding-right");
+        new_size.height += style_size_property(&style, "border-top-width")
+            + style_size_property(&style, "border-bottom-width")
+            + style_size_property(&style, "padding-top")
+            + style_size_property(&style, "padding-bottom");
+    }
 
-    let physical_size = new_size.to_physical(scale_factor);
-    canvas.size().set(physical_size);
+    set_canvas_style_property(raw, "width", &format!("{}px", new_size.width));
+    set_canvas_style_property(raw, "height", &format!("{}px", new_size.height));
+}
 
-    let logical_size = new_size.to_logical::<f64>(scale_factor);
-    set_canvas_style_property(canvas.raw(), "width", &format!("{}px", logical_size.width));
-    set_canvas_style_property(
-        canvas.raw(),
-        "height",
-        &format!("{}px", logical_size.height),
-    );
+/// This function will panic if the element is not inserted in the DOM
+/// or is not a CSS property that represents a size in pixel.
+pub fn style_size_property(style: &CssStyleDeclaration, property: &str) -> f64 {
+    let prop = style
+        .get_property_value(property)
+        .expect("Found invalid property");
+    prop.strip_suffix("px")
+        .expect("Element was not inserted into the DOM or is not a size in pixel")
+        .parse()
+        .expect("CSS property is not a size in pixel")
 }
```

Comment:
`getPropertyValue` doesn't always return a value in `px` - it _usually_ does, but in the special case where `display` is `none` or `contents` it doesn't.

I found this out in https://github.com/rust-windowing/winit/pull/2074/files#r961531983, the summary of which is that `border-*-width` properties always return values in `px`, but the `padding-*` properties might return percentages when `display` is `none` or `contents`.

I looked into how `width` and `height` work here as well, and they work the same way as `padding-*` - when `display` is `none` or `contents`, they might return percentages, or keywords like `min-content`. But otherwise they'll always return a value in `px`.

The solution to this in `ResizeScaleInternal::notify` is pretty simple: just return (0, 0), since that's what `ResizeObserver` reports for elements with `display: none` or `display: contents`. (side-note: I'm not sure if this is actually defined in the spec, which just says to return the element's content area. Content area's a property of the element's box, and I think elements with `display: none` don't have a box, so I'm not sure where that's supposed to come from.)

For `set_canvas_size` it's a bit trickier. We could just do nothing, but I think it'd make more sense to try and make it so that if/when `display` gets set back to something other than `none` or `contents`, it'll have that size. But we can't really do that properly without access to the size of the element's padding and borders.

That also opens up another problem - if the element's padding and/or borders are defined as relative sizes, they can change, meaning that the actual inner size of the canvas can drift from what it was set to if the window changes size.

One solution would be to also set `box-sizing: content-box`, and then not worry about padding and borders. The problem with that is that it also affects `min/max-width/height`, so if those are set it could mess with their meaning. One way to solve that would be to also apply `min-width: unset; max-width: unset; ...` to the canvas, since you could argue that `set_inner_size` is supposed to 100% guarantee that it'll have that size and capping it to a min/max size in CSS would conflict with that; but that might be reaching too far.

Reply:
I'm a bit close to crying.

If we detect `display: none` (or `contents`, which resolves to `none` in the case of the canvas anyway), returning zero seems sensible to me.

> For set_canvas_size it's a bit trickier. We could just do nothing, but I think it'd make more sense to try and make it so that if/when `display` gets set back to something other than `none` or `contents`, it'll have that size.

We should do nothing in that case, `Window::set_inner_size()` should become `request_inner_size()` in the future hopefully anyway, see #2868.

> That also opens up another problem - if the element's padding and/or borders are defined as relative sizes, they can change, meaning that the actual inner size of the canvas can drift from what it was set to if the window changes size.

This is starting to get a bit overkill and we have to draw the line somewhere. As far as I decided in #2859, we won't be setting any CSS properties at all unless requested by `Window::set_inner_size()` (or by `WindowEvent::ScaleFactorChanged::new_inner_size`). Setting the window size is only a request, which we will fulfill to the best of our abilities (including calculating border and padding sizes), but the size we report back might be slightly different, e.g. a pixel off if we are using `devicePixelContentBoxSize`.

The thing we really care about is reporting `WindowEvent::Resized` if the physical size changed and report that as accurate as possible, which I believe is still intact. Even if the element's padding or borders are relative, `ResizeObserver` will still pick up any changes and report them to us.

---

Thank you for the careful review so far @Liamolucko, it was really really helpful.

Reply:
I addressed this as explained above.